Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs

```diff
@@ -81,67 +82,195 @@ public void Initialize(IBuildServerWatcher buildServerWatcher, ISettingsSource c
+                || prev.CommitHashList[0] != curr.CommitHashList[0])
+            {
+                res = true;
+            }
+
+            prev = curr;
+            return res;
+        }
+
         private void ObserveBuilds(DateTime? sinceDate, bool? running, IObserver<BuildInfo> observer, CancellationToken cancellationToken)
         {
+            //Note that 'running' is ignored (attempt to fetch data when updated) 
+            //Similar for 'sinceDate', not supported in Jenkins API
             try
             {
-                var buildUrls = GetBuildUrls(running ?? false);
-                if (buildUrls.All(t => t.IsCanceled))
+                IList<Task<ResponseInfo>> allBuildInfos = new List<Task<ResponseInfo>>();
+                IList<Task<ResponseInfo>> latestBuildInfos = new List<Task<ResponseInfo>>();
+                foreach (var projectUrl in _projectsUrls)
+                {
+                    //Update the build info from the cache and find if (inprogress) jobs may need to be requeried
+                    bool hasInProgress = false;
+                    if (BuildCache.ContainsKey(projectUrl))
+                    {
+                        BuildInfo prevBuild = null;
+                        foreach (var buildInfo in BuildCache[projectUrl].BuildInfo)
+                        {
+                            //Update revision grid from the cache when switching back to this repo
+                            if (_firstQueryAfterInit && DifferFromPrev(buildInfo, ref prevBuild))
+                            {
+                                observer.OnNext(buildInfo);
+                            }
+
+                            //If any build is InProgress, it has to be requeried
+                            //This could be done per build too, probably decreasing load only in special situations
+                            //(Updating all builds individually increases the total load)
+                            if (buildInfo.Status == BuildInfo.BuildStatus.InProgress)
+                            {
+                                hasInProgress = true;
+                            }
+                        }
+                    }
+
+                    if (hasInProgress || !BuildCache.ContainsKey(projectUrl)
+                                   || BuildCache[projectUrl].Timestamp <= 0)
+                    {
+                        //This job must be updated, no need to to check the cache
+                        BuildCache[projectUrl] = new JenkinsCacheInfo();
+                        allBuildInfos.Add(GetBuildInfoTask(projectUrl, true, cancellationToken));
+                    }
+                    else
+                    {
+                        //Check the existing build cache
+                        latestBuildInfos.Add(GetBuildInfoTask(projectUrl, false, cancellationToken));
+                    }
+                }
+                _firstQueryAfterInit = false;
```

Comment:
I would remove _firstQueryAfterInit  and hasInProgress flags and the related logic. We have to continuously query for new builds every ten seconds. Perhaps it is a better idea to limit the fetch size rather than the tree depth (this requires performing two calls). The build numbers are sequential, so we know if the current fetch size is too small or too large and we can adjust it for the next fetch.

Reply:
_firstQueryAfterInit is to display info quickly when switching back. Do you suggest to remove the cache completely or to always update from cache? 
The Apache server I tested to is slow and unstable... 

The fetch size including builds is with the change comparable to the size to fetch the builds before. As the fetch is done first only, then only last build and in progress, the normal load is much lower. 

The separate fetch (and duplicate fetch) caused the instability in the current implementation. A hybrid implementation where all builds are fetched first, then check for last build, fetch new builds, then individual in progress and new builds would decrease network traffic further but I do not feel it is worth it. If still requested, I would prefer to change this later, I suggest this issue to be considered for 2.51

Thanks for reviewing

Reply:
I suggest to always update from cache (after running an async request for the current data)

Reply:
The fetch times for 
`https://builds.apache.org/job/HBase-Flaky-Tests/api/json?depth=1&tree=lastBuild[timestamp]`
and 
`https://builds.apache.org/job/HBase-Flaky-Tests/api/json?depth=1&tree=lastBuild[timestamp],builds[number,result,timestamp,url,actions[lastBuiltRevision[SHA1],totalCount,failCount,skipCount],building,duration]`
are nearly the same -  about 100 milliseconds. I see no point in introducing a special code to handle them separately. Moreover, it makes pointless my proposal of adjusting fetch size.

Reply:
Limiting the request will limit the load on the Jenkins server though, that was my primary intention. Secondly limit processing in ge. But the difference was small enough to not introduce the third level, data per build.